src/qminer/qminer_core.cpp

```diff
@@ -5877,31 +5910,66 @@ void TIndex::IndexTextPos(const int& KeyId, const TStr& TextStr, const uint64& R
     IndexTextPos(KeyId, WordIdV, RecId);
 }
 
-void TIndex::IndexTextPos(const int& KeyId, const TUInt64V& WordIdV, const uint64& RecId) {
+void TIndex::UpdateTextPos(const int& KeyId, const TUInt64V& WordIdV, const uint64& RecId, void (TGix<TQmGixKey, TQmGixItemPos>::*UpdateMethod)(const TQmGixKey&, const TQmGixItemPos&)) {
     // we shouldn't modify read-only index
     QmAssertR(!IsReadOnly(), "Cannot edit read-only index!");
-    // aggregate by word
-    THash<TUInt64, TQmGixItemPos> WordIdPosH;
-    for (int WordIdN = 0; WordIdN < WordIdV.Len(); WordIdN++) {
-        const uint64 WordId = WordIdV[WordIdN];
-        // check if first time we see the word
-        if (!WordIdPosH.IsKey(WordId)) {
-            WordIdPosH.AddDat(WordId, TQmGixItemPos(RecId));
+    // for each word we want to build the TQmGixItemPos that is already sorted from beginning. For that reason we
```

Comment:
This block of code is way too complicated. I would just change `WordIdPosH` to be `THash<TUInt64, TIntSet>`, create static function on `TQmGixItemPos` that can cut word position to whatever the internal modulo is (lets call it `TQmGixItemPos::NormPos(...)` and then the loop becomes something in the following style:

```
for (int WordIdN = 0; WordIdN < WordIdV.Len(); WordIdN++) {
    const uint64 WordId = WordIdV[WordIdN];
    WordIdPosH.AddDat(WordId).AddKey(TQmGixItemPos::NormPos(WordIdN))
}
```